### ISSUE TITLE
reproducibility of paddle2.1.2

### ISSUE THREAD

**Issue by crj1998** [PaddlePaddle/Paddle#36305]
The reproducibility of Paddle2x is ambiguous. The result is various even for fixed seed(random, numpy, paddle).
```
from tqdm import tqdm
import numpy as np
import random
from datetime import date

import paddle
import paddle.nn as nn
# import paddle.nn.functional as F
import paddle.optimizer as optim
import paddle.vision.transforms as transforms
from paddle.vision.datasets import Cifar10
from paddle.io import DataLoader

# Hyper-parameters
class args:
    seed = 0
    lr = 0.1
    epoch_num = 10
    milestones = [5, 7]
    batch_size = 128
    steps = 10
    step_size = 2/255
    epsilon = 8/255

def setup_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    paddle.seed(seed)

def logits_accuracy(logits, labels):
    preds = logits.argmax(axis=-1)
    acc = (preds==labels).astype("float32").sum().item()/labels.shape[0]
    return acc

# ResNet 18 for CIFAR10
class BasicBlock(nn.Layer):
    expansion = 1
    def __init__(self, in_channels, out_channels, stride=1):
        super(BasicBlock, self).__init__()
        self.conv1 = nn.Conv2D(in_channels, out_channels, kernel_size=3, stride=stride, padding=1, bias_attr=False)
        self.bn1 = nn.BatchNorm2D(out_channels)
        self.conv2 = nn.Conv2D(out_channels, out_channels, kernel_size=3, stride=1, padding=1, bias_attr=False)
        self.bn2 = nn.BatchNorm2D(out_channels)
        self.relu = nn.ReLU()

        # shortcut down sample
        if stride != 1 or in_channels != self.expansion*out_channels:
            self.shortcut = nn.Sequential(
                ("conv", nn.Conv2D(in_channels, self.expansion*out_channels, kernel_size=1, stride=stride, padding=0, bias_attr=False)),
                ("bn", nn.BatchNorm2D(self.expansion*out_channels))
            )
        else:
            self.shortcut = nn.Sequential()

    def forward(self, x):
        identity = self.shortcut(x)

        out = self.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))

        out = out + identity
        out = self.relu(out)

        return out

class ResNet(nn.Layer):
    factor = 1.0
    assert factor in [0.25, 0.5, 1.0, 2.0, 4.0]
    _C = (np.array([64, 128, 256, 512])*factor).astype(np.int64)
    def __init__(self, block, num_blocks, num_classes):
        super(ResNet, self).__init__()
        self.channels = self._C[0]
        self.layer0 = nn.Sequential(
            ("conv", nn.Conv2D(in_channels=3, out_channels=self._C[0], kernel_size=3, stride=1, padding=1, bias_attr=False)),
            ("bn", nn.BatchNorm2D(num_features=self._C[0])),
            ("relu", nn.ReLU())
        )

        self.layer1 = self.make_layer(block, self._C[0], num_blocks[0], 1)
        self.layer2 = self.make_layer(block, self._C[1], num_blocks[1], 2)
        self.layer3 = self.make_layer(block, self._C[2], num_blocks[2], 2)
        self.layer4 = self.make_layer(block, self._C[3], num_blocks[3], 2)

        self.avgpool = nn.AdaptiveAvgPool2D((1, 1))
        self.flatten = nn.Flatten(start_axis=1)
        self.fc = nn.Linear(self._C[3]*block.expansion, num_classes)

    def make_layer(self, block, out_channels, block_num, stride):
        layers = [("block0", block(self.channels, out_channels, stride))]
        self.channels = out_channels * block.expansion
        for i in range(block_num - 1):
            layers.append((f"block{i+1}", block(self.channels, out_channels, 1)))
            self.channels = out_channels * block.expansion

        return nn.Sequential(*layers)

    def forward(self, x):
        x = self.layer0(x)     # (N,  64, 16, 16)
        x = self.layer1(x)     # (N,  64, 16, 16)
        x = self.layer2(x)     # (N, 128,  8,  8)
        x = self.layer3(x)     # (N, 256,  4,  4)
        x = self.layer4(x)     # (N, 512,  2,  2)
        x = self.avgpool(x)    # (N, 512,  1,  1)
        x = self.flatten(x)    # (N, 512)
        x = self.fc(x)         # (N, 10)

        return x
    
def resnet18(num_classes):
    return ResNet(BasicBlock, [2, 2, 2, 2], num_classes)


@paddle.set_grad_enabled(True)
def train(epoch, model, dataloader, criterion, optimizer):
    Acc, Loss = 0, 0
    model.train()
    with tqdm(enumerate(dataloader), total=dataloader.__len__(), desc=f"Train Epoch {epoch}") as t:
        for i, (images, labels) in t:
            logits = model(images)
            loss = criterion(logits, labels)

            optimizer.clear_grad()
            loss.backward()
            optimizer.step()

            Loss += loss.item()
            Acc += logits_accuracy(logits.detach(), labels)

            t.set_postfix({"Loss": f"{Loss/(i+1):6.3f}", "Acc": f"{Acc/(i+1):4.2%}"})

    return Acc/(i+1), Loss/(i+1)


@paddle.no_grad()
def test(epoch, model, dataloader):
    Acc = 0
    model.eval()
    with tqdm(enumerate(dataloader), total=dataloader.__len__(), desc="Test") as t:
        for i, (images, labels) in t:
            Acc += logits_accuracy(model(images), labels)
            t.set_postfix({"Acc": f"{Acc/(i+1):4.2%}"})
    return Acc/(i+1)



from visualdl import LogWriter

device = paddle.device.set_device("gpu")
setup_seed(args.seed)

net = resnet18(10)
criterion = nn.CrossEntropyLoss()
scheduler = optim.lr.MultiStepDecay(learning_rate=args.lr, milestones=args.milestones, gamma=0.1, verbose=False)
optimizer = optim.SGD(learning_rate=scheduler, parameters = net.parameters())


train_transform = transforms.Compose([
    transforms.RandomCrop(32, padding=4),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor()
])
test_transform = transforms.Compose([
    transforms.ToTensor()
])
train_set = Cifar10(data_file="../data/data9154/cifar-10-python.tar.gz", mode="train", download=True, transform=train_transform)
test_set = Cifar10(data_file="../data/data9154/cifar-10-python.tar.gz", mode="test", download=True, transform=test_transform)
train_loader = DataLoader(train_set, batch_size=args.batch_size, shuffle=True, drop_last=False, num_workers=2)
test_loader = DataLoader(test_set, batch_size=args.batch_size, shuffle=False, drop_last=False, num_workers=2)


with LogWriter(logdir="./log/"+date.today().strftime('%Y%m%d'), filename_suffix="quarter") as writer:
    for epoch in range(1, args.epoch_num+1):
        train_acc, train_loss = train(epoch, net, train_loader, criterion, optimizer)
        test_acc = test(epoch, net, test_loader)
        scheduler.step()
        writer.add_scalar(tag="train/acc", step=epoch, value=train_acc)
        writer.add_scalar(tag="train/loss", step=epoch, value=train_loss)
        writer.add_scalar(tag="test/acc", step=epoch, value=test_acc)
```

**Comment by GT-ZhangAcer** [comment]
## 由QQ群补充的信息  
<img width="452" alt="图片" src="https://user-images.githubusercontent.com/46156734/136646541-fe80ca80-6620-4d65-8e9e-2a51105af77c.png">

**Comment by gongel** [comment]
你好，已收到问题反馈。可以先尝试设置环境变量：`FLAGS_cudnn_deterministic=True`来进行固定。

**Comment by GT-ZhangAcer** [comment]
我也试了一下，在本地的MAC和AI Studio上进行尝试，本地可以锁死，AI Studio的CPU出现了不能锁死的情况。

**Comment by crj1998** [comment]
> 你好，已收到问题反馈。可以先尝试设置环境变量：`FLAGS_cudnn_deterministic=True`来进行固定。

Yeah, I have tried use `FLAGS_cudnn_deterministic=True python main.py` on AI studio, but the result is still  unrepeatable.

**Comment by gongel** [comment]
> > 你好，已收到问题反馈。可以先尝试设置环境变量：`FLAGS_cudnn_deterministic=True`来进行固定。
> 
> Yeah, I have tried use `FLAGS_cudnn_deterministic=True python main.py` on AI studio, but the result is still unrepeatable.

好的，可以尝试在别的GPU机器上测试。

**Comment by gongel** [comment]
> 我也试了一下，在本地的MAC和AI Studio上进行尝试，本地可以锁死，AI Studio的CPU出现了不能锁死的情况。

一般情况下，CPU机器是可以固定的，GPU可能存在不固定的情况。AI studio待排查。

**Comment by GT-ZhangAcer** [comment]
这个我看QQ群里面对这个需求的人还挺多，这个大概什么时候可以给个结论呢，我也好和用户去讲这个情况

**Comment by gongel** [comment]
Hi all，已确认这是个已知的问题，我们会尽快改进，感谢您提出的宝贵意见。

**Comment by gongel** [comment]
Hi all，更新一下进展，针对ResNet，这边给出以下解决方法，其他的网络还在排查中。
1、export FLAGS_cudnn_deterministic=True 
2、dataloader的number worker 设为0，并关掉shuffle
3、设置 numpy、paddle、random的seed
